src/test/libraryfilesynctest.cpp

```diff
@@ -0,0 +1,711 @@
+#include <gtest/gtest.h>
+
+#include <QFile>
+#include <QFileInfo>
+#include <QTemporaryDir>
+#include <memory>
+
+#include "library/library_prefs.h"
+#include "sources/metadatasource.h"
+#include "test/librarytest.h"
+#include "track/track.h"
+#include "util/fileinfo.h"
+
+namespace {
+
+const auto kTestDir = QDir(QDir::current().absoluteFilePath("src/test/id3-test-data"));
+
+const auto kTestFileWithMetadata = QFileInfo(kTestDir, QStringLiteral("cover-test-jpg.mp3"));
+
+const auto kTestFileWithoutMetadata = QFileInfo(kTestDir, QStringLiteral("empty.mp3"));
+
+enum class FileType {
+    WithMetadata,
+    WithoutMetadata,
+};
+
+void sleepAfterFileLastModifiedUpdated() {
+    // Ensure that all subsequent modification time stamps will be
+    // strictly greater, i.e. strictly monotonic sequence of time
+    // stamps. This grace period seems to be mandatory for the CI
+    // test runs and might need to be adjusted depending on the
+    // results for different platforms.
+    QThread::msleep(1);
+}
+
+/// A temporary file system with a single audio file.
+class TempFileSystem {
+  public:
+    explicit TempFileSystem(const QFileInfo& testFile)
+            : m_fileInfo(m_tempDir.filePath(testFile.fileName())) {
+        createFile(testFile);
+    }
+
+    const mixxx::FileInfo& fileInfo() const {
+        return m_fileInfo;
+    }
+
+    QDateTime fileLastModified() const {
+        return m_fileInfo.toQFile().fileTime(QFileDevice::FileModificationTime);
+    }
+
+    void updateFileLastModified() const {
+        auto file = m_fileInfo.toQFile();
+        const auto oldLastModified = fileLastModified();
+        ASSERT_TRUE(oldLastModified.isValid());
```

Comment:
ASSERT_ macros are not (well) designed to use in subroutines. They are designed to immediately stop the test when used in the test body. Unfortunately they do not abort the whole run like normal asserts, the just return from the function, which is not desired here. You should replace them with EXPECT_ macros, for a deterministic behavior of the test assumptions in the following code. This way you also get rid of the requirement to use void functions in tests.

Reply:
The gtest ASSERT macros are used to very preconditions. These preconditions are "established" in subroutines.

Reply:
Sorry, that is simply not true. 

Please use EXPECT, introduce a proper return value instead of the pointer parameter and both topics are solved.

Reply:
If preconditions are violated tests should be aborted immediately. That's what assertions are for. Please consult the official docs and tell me what I am misunderstanding about it:

http://google.github.io/googletest/reference/assertions.html

Reply:
I think Daniel means that the ASSERT macros only abort the current function, not the whole test (which would be the right thing to do):

> Upon failure, `EXPECT_` macros generate nonfatal failures and allow the current function to continue running, while `ASSERT_` macros generate fatal failures and abort the current function.

But I'm actually now sure how that is supposed to work. I guess the documentation is misleading. Otherwise I have no idea what would happen if the function had a return value?

Reply:
If that is indeed the case, then `EXPECT` would also be wrong and we should make the hlper function return a bool instead, and assert that the output value is true in the actual test case.

Reply:
It was returning a boolean before but that has proven to be tedious to use. I'll check again.

Reply:
Returning a boolean would require a return statement after every EXPECT macro. I was  using VERIFY_AND_DEBUG_ASSERT for this purpose but that only works in debug mode and does not indicate the reason for the failure when compiling with DEBUG_ASSERTIONS_FATAL=OFF. It is not suitable for testing.

Reply:
I am not going to write a new testing framework to overcome the deficiencies of GoogleTest.

Reply:
https://github.com/mixxxdj/mixxx/pull/4628/commits/a4f550ba80c3e2af1058bf5f4bcedf8cd5df1186